src/metabase/api/slack.clj

```diff
@@ -10,20 +11,32 @@
 
 (api/defendpoint PUT "/settings"
   "Update Slack related settings. You must be a superuser to do this."
-  [:as {{slack-app-token :slack-app-token} :body}]
-  {slack-app-token     (s/maybe su/NonBlankString)}
+  [:as {{slack-app-token :slack-app-token, slack-files-channel :slack-files-channel} :body}]
+  {slack-app-token     (s/maybe su/NonBlankString)
+   slack-files-channel (s/maybe su/NonBlankString)}
   (api/check-superuser)
-  (if-not slack-app-token
-    (slack/slack-app-token nil)
-    (try
-      (when-not config/is-test?
-        (when-not (slack/valid-token? slack-app-token)
-          (throw (ex-info (tru "Invalid Slack token.") {:status-code 400}))))
-      ;; Clear the deprecated `slack-token` when setting a new `slack-app-token`
-      (slack/slack-token nil)
-      (slack/slack-app-token slack-app-token)
-      {:ok true}
-      (catch clojure.lang.ExceptionInfo info
-        {:status 400, :body (ex-data info)}))))
+  (try
+    (when (and slack-app-token (not config/is-test?))
+      (when-not (slack/valid-token? slack-app-token)
+        (throw (ex-info (tru "Invalid Slack token.") {:status-code 400}))))
+    (slack/slack-app-token slack-app-token)
+    (when slack-app-token
+      (do
+        (slack/slack-token-valid? true)
```

Comment:
I had to read this line `(slack/slack-token-valid? true)` five times to understand it. This makes me think settings aren't exactly what we want here. An analogy is using three atoms for `slack-app-token`, `slack-token-valid?`, and `slack-files-channel` instead of a single setting or atom for `{:app-token "foo" :valid? true :channel "foo"}`. Not suggesting changing it now but this config spread out over a few different vars feels a bit off.

Reply:
Yeah I agree that having a separate setting to keep track of whether the current token is still valid is not ideal. But I wasn't sure how to implement this better. Are you suggesting having a single setting that contains both the token and whether it's valid?

Maybe I could improve the setting name to make it less confusing to read...I considered `slack-enabled`, since when you get a token error Slack is essentially "disabled" until you save a new, valid token. But that's also a bit iffy 🤷

Reply:
fair enough. And there'd probably be an annoying amount of work to make the admin screens handle the json blob rather than nice fields in it. Oh well. I still think there's some tension with "settings" and "state" that we haven't quite figured out yet but this looks good to me.